### PR TITLE
User status bug fixing...

### PR DESCRIPTION
**Pull request by camilasan**
- [x] address Kevin's last comments in the PR #2505 
- [x] if there is no emoji in the status message (and for the default "Online, offline, etc" ones), the text is a bit off to the right, not left-aligned with name and server link
- [x] the menu icon for the other apps is missing on the right next to the Talk icon, and Talk icon moved over? Unrelated to status, but visible in the screenshots and also in the release it seems
apps

### CODE REVIEW

File: src/gui/tray/Window.qml [hunk 1]
```diff
@@ -403,67 +403,59 @@ Window {
...
                 }
 
                 HeaderButton {
                     id: trayWindowTalkButton
-
                     visible: UserModel.currentUser.serverHasTalk
                     icon.source: "qrc:///client/theme/white/talk-app.svg"
                     onClicked: UserModel.openCurrentAccountTalk()
-
+
```
Comment (FlexW) [inline src/gui/tray/Window.qml:458]: Unnecessary whitespace :)

File: src/gui/userstatus.h [hunk 2]
```diff
@@ -42,23 +42,16 @@ class UserStatus : public QObject
...
     Status stringToEnum(const QString &status) const;
-    
-    // it needs to match the Status enum
-    const QHash<QString, int> _preDefinedStatus{{"online", 0},
-                                               {"dnd", 1}, //DoNotDisturb
-                                               {"away", 2},
-                                               {"offline", 3},
-                                               {"invisible", 4}};
-
+    QString enumToUserString(Status status) const;
```
Comment (allexzander) [inline src/gui/userstatus.h:52]: @camilasan shouldn't it be named just `enumToString` without **user**? Otherwise, the opposite conversion method should also contain **user** and should become `stringToEnum` -> `userStringToEnum`. Otherwise, I can notice inconsistency in naming.

File: src/gui/userstatus.cpp [hunk 3]
```diff
@@ -30,16 +30,37 @@ Q_LOGGING_CATEGORY(lcUserStatus, "nextcloud.gui.userstatus", QtInfoMsg)
...
 UserStatus::UserStatus(QObject *parent)
     : QObject(parent)
-    , _message("")
 {
 }
 
 UserStatus::Status UserStatus::stringToEnum(const QString &status) const 
 {
+    // it needs to match the Status enum
+    const QHash<QString, Status> preDefinedStatus{{"online", Online},
```
Comment (FlexW) [inline src/gui/userstatus.cpp:39]: Maybe there was a reason for making that a local variable that I overlooked, but why is not a static local variable?

File: src/gui/userstatus.cpp [hunk 4]
```diff
@@ -30,16 +30,37 @@ Q_LOGGING_CATEGORY(lcUserStatus, "nextcloud.gui.userstatus", QtInfoMsg)
...
     // it matches _preDefinedStatus, otherwise the default is online (0)
-    const auto statusEnum = _preDefinedStatus.value(status.isEmpty()? "online" : status.toLower(), 0);
-    return static_cast<Status>(statusEnum);
+    const auto statusKey = status.isEmpty() ? QStringLiteral("online") : status.toLower();
+    return preDefinedStatus.value(statusKey, Online);
+}
+
+QString UserStatus::enumToUserString(Status status) const 
+{
+    switch (status) {
```
Comment (allexzander) [inline src/gui/userstatus.cpp:53]: @camilasan I am still wondering why we can't just use `QString statusStrings[Status::Count];` and then just `return statusStrings[status];` ? :)

File: src/gui/userstatus.cpp [hunk 5]
```diff
@@ -53,10 +74,9 @@ void UserStatus::fetchUserStatus(AccountPtr account)
     _job->start();
 }
 
-void UserStatus::slotFetchUserStatusFinished(const QJsonDocument &json, const int statusCode)
+void UserStatus::slotFetchUserStatusFinished(const QJsonDocument &json, int statusCode)
 {
-    const QJsonObject defaultValues
-    {
+    const QJsonObject defaultValues {
```
Comment (FlexW) [inline src/gui/userstatus.cpp:79]: Same question as above :)

File: src/gui/userstatus.cpp (hunk 3 above)
Comment (allexzander) [inline src/gui/userstatus.cpp:39]: @camilasan And, I think, this could just as well be implemented via `if`/`else if` for strings:
`if (statusString == "dnd") { return DoNotDisturb; }` and so on... Or, even, `std::find` on array of enum strings as proposed in enumToUserString. The search is linear in that case, but, the data size is so small, so, I can hardly see any significant benefit of using `QHash` here.

File: src/gui/tray/Window.qml [hunk 6]
```diff
@@ -403,67 +403,59 @@ Window {
...
-                    sourceSize.width: Style.folderStateIndicatorSize
-                    sourceSize.height: Style.folderStateIndicatorSize
-
-                    Accessible.role: Accessible.Indicator
-                    Accessible.name: UserModel.isUserConnected(UserModel.currentUserId()) ? qsTr("Connected") : qsTr("Disconnected")
                 }
 
                 HeaderButton {
                     id: trayWindowTalkButton
-
```
Comment (allexzander) [inline src/gui/tray/Window.qml:462]: @camilasan Are these changes necessary?

File: src/gui/tray/Window.qml [hunk 7]
```diff
@@ -403,67 +403,59 @@ Window {
...
                     visible: UserModel.currentUser.serverHasTalk
                     icon.source: "qrc:///client/theme/white/talk-app.svg"
                     onClicked: UserModel.openCurrentAccountTalk()
-
```
Comment (allexzander) [inline src/gui/tray/Window.qml:466]: @camilasan Are these changes necessary?

File: src/gui/tray/Window.qml (hunk 6 above)
Comment (allexzander) [inline src/gui/tray/Window.qml:462]: @camilasan Are these changes necessary?

File: src/gui/userstatus.cpp [hunk 8]
```diff
@@ -30,16 +30,37 @@ Q_LOGGING_CATEGORY(lcUserStatus, "nextcloud.gui.userstatus", QtInfoMsg)
...
+    
     // api should return invisible, dnd,... toLower() it is to make sure 
     // it matches _preDefinedStatus, otherwise the default is online (0)
-    const auto statusEnum = _preDefinedStatus.value(status.isEmpty()? "online" : status.toLower(), 0);
-    return static_cast<Status>(statusEnum);
+    const auto statusKey = status.isEmpty() ? QStringLiteral("online") : status.toLower();
+    return preDefinedStatus.value(statusKey, Online);
+}
+
+QString UserStatus::enumToUserString(Status status) const
```
Comment (FlexW) [inline src/gui/userstatus.cpp:51]: Could also made into a static function of the cpp file. This way it would not "pollute" the class definition. A other question that I have is: Is the name of the function not to generic? For example I would choose `statusEnumToUserString()`.

File: src/gui/tray/Window.qml [hunk 9]
```diff
@@ -403,67 +403,59 @@ Window {
...
+                        visible: UserModel.currentUser.hasLocalFolder
+                        icon.source: "qrc:///client/theme/white/folder.svg"
+                        onClicked: UserModel.openCurrentAccountLocalFolder()
+                    }
+                                             
+                    Rectangle {
+                        id: folderStateIndicatorBackground
+                        width: Style.folderStateIndicatorSize
+                        height: width
+                        anchors.top: openLocalFolderButton.verticalCenter
```
Comment (mgallien) [inline src/gui/tray/Window.qml:427]: I believe that anchors and being inside a layout is undefined behavior and should be avoided.
You could have only the HeaderButton in the layout and the Rectangle as a child of it set at the desired position and size.

I am getting lots of warning about that in the log

> QML Rectangle: Detected anchors on an item that is managed by a layout. This is undefined behavior; use Layout.alignment instead.

File: src/gui/userstatus.cpp (hunk 3 above)
Comment (er-vin) [reply src/gui/userstatus.cpp:39]: It's kind of a micro-optimization really, but adding static can't hurt here really.

File: src/gui/tray/Window.qml [hunk 10]
```diff
@@ -403,67 +403,59 @@ Window {
...
-                    visible: UserModel.currentUser.hasLocalFolder
-                    icon.source: "qrc:///client/theme/white/folder.svg"
-                    onClicked: UserModel.openCurrentAccountLocalFolder()
+                    Image {
+                        id: folderStateIndicator
+                        source: UserModel.isUserConnected(UserModel.currentUserId)
+                                ? Style.stateOnlineImageSource
+                                : Style.stateOfflineImageSource
+                        cache: false
+                        x: folderStateIndicatorBackground.x
```
Comment (mgallien) [inline src/gui/tray/Window.qml:439]: Explicit setting of x, y, width and height is discouraged in the Layout documentation to avoid running into binding loops.
You may also put it as a child of the HeaderButton.

File: src/gui/userstatus.cpp (hunk 3 above)
Comment (er-vin) [reply src/gui/userstatus.cpp:39]: I don't think it's an algorithmic performance thing, it's mainly a question of style I'd say. It's a table look up so it makes sense to have a table to look into. :-)

IME there's less cognitive load in reading the table look up version months from now than the if / else if version + factor in the lower case + factor in the missing status.

File: src/gui/tray/Window.qml [hunk 11]
```diff
@@ -403,67 +403,59 @@ Window {
...
-                    Layout.fillWidth: true
-                }
-
-                HeaderButton {
-                    id: openLocalFolderButton
+                RowLayout {
+                    id: openLocalFolderRowLayout
+                    height: Style.trayWindowHeaderHeight
+                    width: height
+                    anchors.left: accountControlRowLayout.right
```
Comment (mgallien) [inline src/gui/tray/Window.qml:410]: anchors and Layout are incompatible. Please see my other comments.

File: src/gui/tray/Window.qml [hunk 12]
```diff
@@ -403,67 +403,59 @@ Window {
...
-                Item {
-                    id: trayWindowHeaderSpacer
-                    Layout.fillWidth: true
-                }
-
-                HeaderButton {
-                    id: openLocalFolderButton
+                RowLayout {
+                    id: openLocalFolderRowLayout
+                    height: Style.trayWindowHeaderHeight
```
Comment (mgallien) [inline src/gui/tray/Window.qml:408]: This is not necessary and is discouraged. Please use only implicitWidth or Layout.preferredWidth .

File: src/gui/userstatus.cpp (hunk 8 above)
Comment (er-vin) [reply src/gui/userstatus.cpp:51]: static functions are so like C++98 ;-)

But the general comment applies, it'd be better as a free function in an anonymous namespace at the beginning of the file.

File: src/gui/userstatus.cpp (hunk 4 above)
Comment (er-vin) [reply src/gui/userstatus.cpp:53]: Because you'd be over-thinking against the enum as it is today. Add a new enum value in the middle and then you got a subtle bug which is not necessarily easy to track down as the string can travel quite a bit until it's consumption point.

Using enum values as indices for arrays is a dangerous business in maintenance. There's a reason for enum class. ;-)

Comment (allexzander) [reply src/gui/userstatus.cpp:53]: @er-vin Alrighty then. Never thought about that, but,  it should also be simple to check that the enum index is within bounds and has a corresponding `QString` assigned to it. Now, if we are talking about `enum class`, why not make that `enum Status` an `enum class Status` instead, @camilasan ? :)

**Comment by er-vin** [comment]
> @er-vin Alrighty then. Never thought about that, but,  it should also be
> simple to check that the enum index is within bounds and has a
> corresponding `QString` assigned to it.

Well, that's why I mentioned "in the middle". Typically sometimes you'd have 
something which is still within bounds but without the message you expect. Of 
course in other cases you'd have something outside of bounds but this is 
easier to catch.

> Now, if we are talking about `enum class`, why not make that `enum Status`
> an `enum class Status` instead, @camilasan ? :)

Yes good point. Thought I mentioned it in the past but probably forgot.

**Comment by allexzander** [comment]
> Well, that's why I mentioned "in the middle". Typically sometimes you'd have something which is still within bounds but without the message you expect. Of course in other cases, you'd have something outside of bounds but this is easier to catch.
@er-vin In one of my past jobs, we were using a command-line tool that would generate all that's needed for enum (array with strings and conversion functions). So, adding an enum index in the middle would just be a simple insertion into an enum template file which is then fed to a command-line tool during the build :)

**Comment by er-vin** [comment]
> @er-vin In one of my past jobs, we were using a command-line tool that would
> generate all that's needed for enum (array with strings and conversion
> functions). So, adding an enum index in the middle would just be a simple
> insertion into an enum template file which is then fed to a command-line
> tool during the build :)

Yes, tooling is a way to escape this. We got that for free in some way with 
Q_ENUM. The difficulty here is the translated strings, Q_ENUM won't help you 
for this and so you're back to square one...

File: src/gui/userstatus.h (hunk 2 above)
Comment (camilasan) [reply src/gui/userstatus.h:52]: Kevin used this name as suggestion in the previous PR...

File: src/gui/userstatus.cpp (hunk 4 above)
Comment (camilasan) [reply src/gui/userstatus.cpp:53]: The response from the server always send e.g. "dnd" and "online" and I am capitalizing the string. Also following Kevin's suggestion in the previous PR. Or I am missing something here...

File: src/gui/tray/Window.qml (hunk 7 above)
Comment (camilasan) [reply src/gui/tray/Window.qml:466]: No :)

File: src/gui/userstatus.cpp (hunk 3 above)
Comment (camilasan) [reply src/gui/userstatus.cpp:39]: See Kevin's comments in the previous PR #2505.

File: src/gui/userstatus.cpp (hunk 4 above)
Comment (camilasan) [reply src/gui/userstatus.cpp:53]: Somehow I missed Kevin's reply.
ok, enum Class it is then.

File: src/gui/userstatus.cpp [hunk 13]
```diff
@@ -66,19 +86,15 @@ void UserStatus::slotFetchUserStatusFinished(const QJsonDocument &json, const in
...
-    
-    // to display it to the user like 'Invisible' instead of 'invisible'
-    statusString.replace(0, 1, statusString.at(0).toUpper());  
 
-    const auto visibleStatusText = message.isEmpty()
-                                ? _status == DoNotDisturb? tr("Do not disturb") 
-                                                : tr(qPrintable(statusString))
-                                : message;
+    _status = stringToEnum(statusString);
+    statusString = enumToString(_status);
```
Comment (allexzander) [inline src/gui/userstatus.cpp:96]: @camilasan IMHO, this line is redundant. You can just write:
`const auto visibleStatusText = message.isEmpty() ? enumToString(_status) : message;` after all. It made me think more when reading this, because, I see this `statusString` is assigned with value before, and then, it is assigned with a new value right after it :)

You may shorter it even more if you just write:
```
_status = stringToEnum(retrievedData.value("status").toString());
const auto visibleStatusText = message.isEmpty() ? enumToString(_status) : message;
```

So, this intermediate variable `statusString` is not needed really.

**Review by allexzander — APPROVED** [review]
Looks fine. Let's fix QML warnings in 3.2.1.

**Comment by er-vin** [comment]
> Looks fine. Let's fix QML warnings in 3.2.1.

This + move stringToEnum and enumToString in an anonymous namespace in the cpp 
file. From the other comments I think that one fell through the cracks.
